File: frontend/src/domain/pace.test.ts
import { describe, it, expect } from 'vitest'
import { computeRunnerPace, projectArrival, formatPace } from './pace'
import type { Checkpoint, CheckpointLog, Runner } from './types'

const makeRunner = (overrides: Partial<Runner> = {}): Runner => ({
  ID: 1,
  RaceID: 1,
  BibNumber: 100,
  FirstName: 'Alice',
  LastName: 'Smith',
  SortOrder: 1,
  Status: 'ACTIVE',
  CreatedAt: '',
  UpdatedAt: '',
  ...overrides,
})

const makeCheckpoint = (id: number, order: number, dist: number | null): Checkpoint => ({
  ID: id,
  RaceID: 1,
  Code: `AS${id}`,
  DisplayName: `Aid Station ${id}`,
  DisplayOrder: order,
  DistanceFromStart: dist,
  CreatedAt: '',
})

const makeLog = (runnerID: number, checkpointID: number, recordedAt: string): CheckpointLog => ({
  ID: checkpointID,
  RunnerID: runnerID,
  CheckpointID: checkpointID,
  RecordedAt: recordedAt,
  Source: 'MANUAL',
  RawMessage: '',
  CreatedAt: recordedAt,
})

describe('computeRunnerPace', () => {
  it('returns nulls for DNS runner', () => {
    const result = computeRunnerPace(makeRunner({ Status: 'DNS' }), [], [])
    expect(result).toEqual({ paceMinPerMile: null, lastLoggedDist: null, lastLoggedAt: null })
  })

  it('returns nulls for DNF runner', () => {
    const result = computeRunnerPace(makeRunner({ Status: 'DNF' }), [], [])
    expect(result).toEqual({ paceMinPerMile: null, lastLoggedDist: null, lastLoggedAt: null })
  })

  it('returns nulls for MOVED runner', () => {
    const result = computeRunnerPace(makeRunner({ Status: 'MOVED' }), [], [])
    expect(result).toEqual({ paceMinPerMile: null, lastLoggedDist: null, lastLoggedAt: null })
  })

  it('returns nulls for FINISHED runner', () => {
    const result = computeRunnerPace(makeRunner({ Status: 'FINISHED' }), [], [])
    expect(result).toEqual({ paceMinPerMile: null, lastLoggedDist: null, lastLoggedAt: null })
  })

  it('returns nulls when runner has no logs', () => {
    const cps = [makeCheckpoint(1, 1, 10.0)]
    const result = computeRunnerPace(makeRunner(), cps, [])
    expect(result).toEqual({ paceMinPerMile: null, lastLoggedDist: null, lastLoggedAt: null })
  })

  it('returns null pace but last position when only one distance checkpoint logged', () => {
    const cps = [makeCheckpoint(1, 1, 10.0), makeCheckpoint(2, 2, 20.0)]
    const logs = [makeLog(1, 1, '2026-06-14T10:00:00Z')]
    const result = computeRunnerPace(makeRunner(), cps, logs)
    expect(result.paceMinPerMile).toBeNull()
    expect(result.lastLoggedDist).toBe(10.0)
    expect(result.lastLoggedAt).toEqual(new Date('2026-06-14T10:00:00Z'))
  })

  it('computes pace from two distance-tagged checkpoints', () => {
    const cps = [makeCheckpoint(1, 1, 10.0), makeCheckpoint(2, 2, 20.0)]
    // 60 minutes between checkpoints, 10 miles apart → 6 min/mi
    const logs = [makeLog(1, 1, '2026-06-14T10:00:00Z'), makeLog(1, 2, '2026-06-14T11:00:00Z')]
    const result = computeRunnerPace(makeRunner(), cps, logs)
    expect(result.paceMinPerMile).toBeCloseTo(6.0)
    expect(result.lastLoggedDist).toBe(20.0)
    expect(result.lastLoggedAt).toEqual(new Date('2026-06-14T11:00:00Z'))
  })

  it('uses the last two distance checkpoints when more than two are logged', () => {
    const cps = [makeCheckpoint(1, 1, 5.0), makeCheckpoint(2, 2, 15.0), makeCheckpoint(3, 3, 25.0)]
    // Between cp2 and cp3: 10 miles in 90 min → 9 min/mi
    const logs = [
      makeLog(1, 1, '2026-06-14T09:00:00Z'),
      makeLog(1, 2, '2026-06-14T10:00:00Z'),
      makeLog(1, 3, '2026-06-14T11:30:00Z'),
    ]
    const result = computeRunnerPace(makeRunner(), cps, logs)
    expect(result.paceMinPerMile).toBeCloseTo(9.0)
  })

  it('skips checkpoints with null distance', () => {
    const cps = [
      makeCheckpoint(1, 1, 10.0),
      makeCheckpoint(2, 2, null), // no distance
      makeCheckpoint(3, 3, 20.0),
    ]
    const logs = [
      makeLog(1, 1, '2026-06-14T10:00:00Z'),
      makeLog(1, 2, '2026-06-14T10:30:00Z'),
      makeLog(1, 3, '2026-06-14T11:00:00Z'),
    ]
    // Between cp1 and cp3: 10 miles in 60 min → 6 min/mi
    const result = computeRunnerPace(makeRunner(), cps, logs)
    expect(result.paceMinPerMile).toBeCloseTo(6.0)
  })

  it('returns null pace when distance delta is zero', () => {
    const cps = [makeCheckpoint(1, 1, 10.0), makeCheckpoint(2, 2, 10.0)]
    const logs = [makeLog(1, 1, '2026-06-14T10:00:00Z'), makeLog(1, 2, '2026-06-14T11:00:00Z')]
    const result = computeRunnerPace(makeRunner(), cps, logs)
    expect(result.paceMinPerMile).toBeNull()
  })

  it('ignores logs for other runners', () => {
    const cps = [makeCheckpoint(1, 1, 10.0), makeCheckpoint(2, 2, 20.0)]
    const logs = [
      makeLog(99, 1, '2026-06-14T10:00:00Z'), // different runner
      makeLog(99, 2, '2026-06-14T11:00:00Z'),
    ]
    const result = computeRunnerPace(makeRunner({ ID: 1 }), cps, logs)
    expect(result.paceMinPerMile).toBeNull()
  })
})

describe('projectArrival', () => {
  const basePace = {
    paceMinPerMile: 6.0,
    lastLoggedDist: 10.0,
    lastLoggedAt: new Date('2026-06-14T10:00:00Z'),
  }

  it('returns null when pace is null', () => {
    expect(projectArrival({ ...basePace, paceMinPerMile: null }, 20)).toBeNull()
  })

  it('returns null when lastLoggedDist is null', () => {
    expect(projectArrival({ ...basePace, lastLoggedDist: null }, 20)).toBeNull()
  })

  it('returns null when lastLoggedAt is null', () => {
    expect(projectArrival({ ...basePace, lastLoggedAt: null }, 20)).toBeNull()
  })

  it('returns null when target is not ahead of last logged position', () => {
    expect(projectArrival(basePace, 10)).toBeNull()
    expect(projectArrival(basePace, 5)).toBeNull()
  })

  it('projects arrival correctly: 10mi at 6min/mi = +60min', () => {
    const result = projectArrival(basePace, 20)
    expect(result).toEqual(new Date('2026-06-14T11:00:00Z'))
  })

  it('projects fractional distances correctly', () => {
    const result = projectArrival(basePace, 15) // 5 miles at 6 min/mi = 30 min
    expect(result).toEqual(new Date('2026-06-14T10:30:00Z'))
  })
})

describe('formatPace', () => {
  it('formats whole minutes', () => {
    expect(formatPace(8)).toBe('8:00 /mi')
  })

  it('formats minutes with seconds', () => {
    expect(formatPace(8.5)).toBe('8:30 /mi')
  })

  it('pads seconds with leading zero', () => {
    expect(formatPace(6.1)).toBe('6:06 /mi')
  })

  it('formats sub-minute pace (edge case)', () => {
    expect(formatPace(0.5)).toBe('0:30 /mi')
  })
})

File: backend/internal/domain/pace/pace_test.go
package pace_test

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/kevinball/ares-bib-logger/backend/internal/domain/entity"
	"github.com/kevinball/ares-bib-logger/backend/internal/domain/pace"
)

func makeRunner(status entity.RunnerStatus) entity.Runner {
	return entity.Runner{ID: 1, RaceID: 1, BibNumber: 100, Status: status}
}

func makeCheckpoint(id, order int, dist *float64) entity.Checkpoint {
	return entity.Checkpoint{ID: id, RaceID: 1, DisplayOrder: order, DistanceFromStart: dist}
}

func makeLog(runnerID, checkpointID int, recordedAt time.Time) entity.CheckpointLog {
	return entity.CheckpointLog{RunnerID: runnerID, CheckpointID: checkpointID, RecordedAt: recordedAt}
}

func dist(v float64) *float64 { return &v }

func mustParse(s string) time.Time {
	t, err := time.Parse(time.RFC3339, s)
	if err != nil {
		panic(err)
	}
	return t
}

func TestComputeRunnerPace_TerminalStatusesReturnEmpty(t *testing.T) {
	for _, status := range []entity.RunnerStatus{
		entity.StatusDNS, entity.StatusDNF, entity.StatusMoved, entity.StatusFinished,
	} {
		result := pace.ComputeRunnerPace(makeRunner(status), nil, nil)
		assert.Nil(t, result.PaceMinPerMile, "status %s", status)
		assert.Nil(t, result.LastLoggedDist, "status %s", status)
		assert.Nil(t, result.LastLoggedAt, "status %s", status)
	}
}

func TestComputeRunnerPace_NoLogsReturnsEmpty(t *testing.T) {
	cps := []entity.Checkpoint{makeCheckpoint(1, 1, dist(10.0))}
	result := pace.ComputeRunnerPace(makeRunner(entity.StatusActive), cps, nil)
	assert.Nil(t, result.PaceMinPerMile)
	assert.Nil(t, result.LastLoggedDist)
}

func TestComputeRunnerPace_OneDistanceCheckpoint_NoPaceButLastPosition(t *testing.T) {
	cps := []entity.Checkpoint{makeCheckpoint(1, 1, dist(10.0)), makeCheckpoint(2, 2, dist(20.0))}
	at := mustParse("2026-06-14T10:00:00Z")
	logs := []entity.CheckpointLog{makeLog(1, 1, at)}

	result := pace.ComputeRunnerPace(makeRunner(entity.StatusActive), cps, logs)
	assert.Nil(t, result.PaceMinPerMile)
	require.NotNil(t, result.LastLoggedDist)
	assert.Equal(t, 10.0, *result.LastLoggedDist)
	require.NotNil(t, result.LastLoggedAt)
	assert.True(t, at.Equal(*result.LastLoggedAt))
}

func TestComputeRunnerPace_TwoDistanceCheckpoints(t *testing.T) {
	cps := []entity.Checkpoint{makeCheckpoint(1, 1, dist(10.0)), makeCheckpoint(2, 2, dist(20.0))}
	// 60 minutes between checkpoints, 10 miles apart -> 6 min/mi
	logs := []entity.CheckpointLog{
		makeLog(1, 1, mustParse("2026-06-14T10:00:00Z")),
		makeLog(1, 2, mustParse("2026-06-14T11:00:00Z")),
	}

	result := pace.ComputeRunnerPace(makeRunner(entity.StatusActive), cps, logs)
	require.NotNil(t, result.PaceMinPerMile)
	assert.InDelta(t, 6.0, *result.PaceMinPerMile, 0.001)
	assert.Equal(t, 20.0, *result.LastLoggedDist)
}

func TestComputeRunnerPace_UsesLastTwoWhenMoreThanTwoLogged(t *testing.T) {
	cps := []entity.Checkpoint{
		makeCheckpoint(1, 1, dist(5.0)),
		makeCheckpoint(2, 2, dist(15.0)),
		makeCheckpoint(3, 3, dist(25.0)),
	}
	// Between cp2 and cp3: 10 miles in 90 min -> 9 min/mi
	logs := []entity.CheckpointLog{
		makeLog(1, 1, mustParse("2026-06-14T09:00:00Z")),
		makeLog(1, 2, mustParse("2026-06-14T10:00:00Z")),
		makeLog(1, 3, mustParse("2026-06-14T11:30:00Z")),
	}

	result := pace.ComputeRunnerPace(makeRunner(entity.StatusActive), cps, logs)
	require.NotNil(t, result.PaceMinPerMile)
	assert.InDelta(t, 9.0, *result.PaceMinPerMile, 0.001)
}

func TestComputeRunnerPace_SkipsCheckpointsWithNilDistance(t *testing.T) {
	cps := []entity.Checkpoint{
		makeCheckpoint(1, 1, dist(10.0)),
		makeCheckpoint(2, 2, nil),
		makeCheckpoint(3, 3, dist(20.0)),
	}
	logs := []entity.CheckpointLog{
		makeLog(1, 1, mustParse("2026-06-14T10:00:00Z")),
		makeLog(1, 2, mustParse("2026-06-14T10:30:00Z")),
		makeLog(1, 3, mustParse("2026-06-14T11:00:00Z")),
	}

	result := pace.ComputeRunnerPace(makeRunner(entity.StatusActive), cps, logs)
	require.NotNil(t, result.PaceMinPerMile)
	assert.InDelta(t, 6.0, *result.PaceMinPerMile, 0.001)
}

func TestComputeRunnerPace_ZeroDistanceDeltaReturnsNilPace(t *testing.T) {
	cps := []entity.Checkpoint{makeCheckpoint(1, 1, dist(10.0)), makeCheckpoint(2, 2, dist(10.0))}
	logs := []entity.CheckpointLog{
		makeLog(1, 1, mustParse("2026-06-14T10:00:00Z")),
		makeLog(1, 2, mustParse("2026-06-14T11:00:00Z")),
	}

	result := pace.ComputeRunnerPace(makeRunner(entity.StatusActive), cps, logs)
	assert.Nil(t, result.PaceMinPerMile)
}

func TestComputeRunnerPace_IgnoresOtherRunnersLogs(t *testing.T) {
	cps := []entity.Checkpoint{makeCheckpoint(1, 1, dist(10.0)), makeCheckpoint(2, 2, dist(20.0))}
	logs := []entity.CheckpointLog{
		makeLog(99, 1, mustParse("2026-06-14T10:00:00Z")),
		makeLog(99, 2, mustParse("2026-06-14T11:00:00Z")),
	}

	result := pace.ComputeRunnerPace(makeRunner(entity.StatusActive), cps, logs)
	assert.Nil(t, result.PaceMinPerMile)
}

func TestProjectArrival(t *testing.T) {
	p6 := 6.0
	d10 := 10.0
	at := mustParse("2026-06-14T10:00:00Z")
	basePace := pace.RunnerPace{PaceMinPerMile: &p6, LastLoggedDist: &d10, LastLoggedAt: &at}

	t.Run("nil pace", func(t *testing.T) {
		p := basePace
		p.PaceMinPerMile = nil
		assert.Nil(t, pace.ProjectArrival(p, 20))
	})

	t.Run("nil last dist", func(t *testing.T) {
		p := basePace
		p.LastLoggedDist = nil
		assert.Nil(t, pace.ProjectArrival(p, 20))
	})

	t.Run("nil last at", func(t *testing.T) {
		p := basePace
		p.LastLoggedAt = nil
		assert.Nil(t, pace.ProjectArrival(p, 20))
	})

	t.Run("target not ahead of last position", func(t *testing.T) {
		assert.Nil(t, pace.ProjectArrival(basePace, 10))
	})

	t.Run("projects arrival", func(t *testing.T) {
		got := pace.ProjectArrival(basePace, 20)
		require.NotNil(t, got)
		want := mustParse("2026-06-14T11:00:00Z")
		assert.True(t, want.Equal(*got), "want %v got %v", want, *got)
	})
}

func TestFormatPace(t *testing.T) {
	assert.Equal(t, "6:00 /mi", pace.FormatPace(6.0))
	assert.Equal(t, "12:30 /mi", pace.FormatPace(12.5))
}

func TestLastLoggedCheckpoint_NoLogsReturnsFalse(t *testing.T) {
	_, _, ok := pace.LastLoggedCheckpoint(nil, nil, 1)
	assert.False(t, ok)
}

func TestLastLoggedCheckpoint_ReturnsHighestDisplayOrderRegardlessOfStatus(t *testing.T) {
	cps := []entity.Checkpoint{
		makeCheckpoint(1, 1, dist(5.0)),
		makeCheckpoint(2, 2, dist(15.0)),
	}
	at := mustParse("2026-06-14T10:00:00Z")
	logs := []entity.CheckpointLog{
		makeLog(1, 1, at),
		makeLog(1, 2, at.Add(30*time.Minute)),
	}

	cp, log, ok := pace.LastLoggedCheckpoint(cps, logs, 1)
	require.True(t, ok)
	assert.Equal(t, 2, cp.ID)
	assert.Equal(t, 2, log.CheckpointID)
}

func TestLastLoggedCheckpoint_AvailableForTerminalStatuses(t *testing.T) {
	// Even though ComputeRunnerPace returns empty for DNF runners,
	// LastLoggedCheckpoint must still report their last known station.
	cps := []entity.Checkpoint{makeCheckpoint(1, 1, dist(5.0))}
	logs := []entity.CheckpointLog{makeLog(1, 1, mustParse("2026-06-14T10:00:00Z"))}

	dnfResult := pace.ComputeRunnerPace(makeRunner(entity.StatusDNF), cps, logs)
	assert.Nil(t, dnfResult.LastLoggedDist, "ComputeRunnerPace should stay empty for DNF")

	cp, _, ok := pace.LastLoggedCheckpoint(cps, logs, 1)
	require.True(t, ok)
	assert.Equal(t, 1, cp.ID)
}
